File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {

        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> current = new ArrayList<Integer>();
        int currIndex = 0;

        getSubSets(current, currIndex, res, nums);


        return res;
    }

    public void getSubSets(List<Integer> current, int currIndex, List<List<Integer>> res, int[] nums){

        if(currIndex == nums.length){
            res.add(new ArrayList<Integer>(current));
            return;
        }

        current.add(nums[currIndex]);
        getSubSets(current, currIndex+1, res, nums);
        current.remove(current.size()-1);
        getSubSets(current, currIndex+1, res, nums);
    }
}

File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:

        ##Time Compleixty : O(N^2)
        ##Space Complexity : O(N)

        res = []

        def getSubSets(current, currIndex):

            if currIndex == len(nums):
                res.append(current.copy())
                return

            current.append(nums[currIndex])
            getSubSets(current, currIndex+1) 
            current.pop()
            getSubSets(current, currIndex+1)           

        getSubSets([], 0)


        return res